pkg/fanal/analyzer/language/python/poetry/poetry.go

```diff
@@ -104,45 +104,46 @@ func (a poetryAnalyzer) mergePyProject(fsys fs.FS, dir string, app *types.Applic
 		return xerrors.Errorf("unable to parse %s: %w", path, err)
 	}
 
-	// Identify the direct/transitive dependencies
+	prodRootDeps := project.Tool.Poetry.Dependencies
+	directDeps := prodRootDeps.Union(getDevDeps(project))
+	prodDeps := getProdPackages(app, prodRootDeps)
+
+	// Identify the direct/transitive/dev dependencies
 	for i, pkg := range app.Packages {
-		if project.Tool.Poetry.Dependencies.Contains(pkg.Name) {
+		app.Packages[i].Dev = !prodDeps.Contains(pkg.ID)
+		if directDeps.Contains(pkg.Name) {
 			app.Packages[i].Relationship = types.RelationshipDirect
 		} else {
 			app.Packages[i].Indirect = true
 			app.Packages[i].Relationship = types.RelationshipIndirect
 		}
 	}
-
-	filterProdPackages(project, app)
 	return nil
 }
 
-func filterProdPackages(project pyproject.PyProject, app *types.Application) {
+func getDevDeps(project pyproject.PyProject) set.Set[string] {
+	deps := set.New[string]()
+	for _, groupDeps := range project.Tool.Poetry.Groups {
+		deps.Append(groupDeps.Dependencies.Items()...)
+	}
+	return deps
+}
```

Comment:
I suggest 2 changes:
1. rename function (see https://go.dev/doc/effective_go#Getters)
2. move `project.Tool.Poetry.Dependencies` into function.
```suggestion
func directDeps(project pyproject.PyProject) set.Set[string] {
	deps := project.Tool.Poetry.Dependencies
	for _, groupDeps := range project.Tool.Poetry.Groups {
		deps.Append(groupDeps.Dependencies.Items()...)
	}
	return deps
}
```